src/hotspot/share/services/heapDumper.cpp

```diff
@@ -608,20 +594,294 @@ void DumpWriter::start_sub_record(u1 tag, u4 len) {
+  void writer_loop()                    { _backend.thread_loop(false); }
+  // Called when finish to release the threads.
+  virtual void deactivate()             { flush(); _backend.deactivate(); }
+  // Get the backend pointer, used by parallel dump writer.
+  CompressionBackend* backend_ptr() { return &_backend; }
+
+};
+
+// Check for error after constructing the object and destroy it in case of an error.
+DumpWriter::DumpWriter(AbstractWriter* writer, AbstractCompressor* compressor) :
+  AbstractDumpWriter(),
+  _backend(writer, compressor, io_buffer_max_size, io_buffer_max_waste) {
+  flush();
+}
+
+// flush any buffered bytes to the file
+void DumpWriter::flush(bool force) {
+  _backend.get_new_buffer(&_buffer, &_pos, &_size, force);
+}
+
+// Buffer queue used for parallel dump.
+struct ParWriterBufferQueueElem {
+  char* _buffer;
+  size_t _used;
+  ParWriterBufferQueueElem* _next;
+};
+
+class ParWriterBufferQueue : public CHeapObj<mtInternal> {
+ private:
+  ParWriterBufferQueueElem* _head;
+  ParWriterBufferQueueElem* _tail;
+  uint _length;
+ public:
+  ParWriterBufferQueue() : _head(NULL), _tail(NULL), _length(0) { }
+
+  void enqueue(ParWriterBufferQueueElem* entry) {
+    if (_head == NULL) {
+      assert(is_empty() && _tail == NULL, "Sanity check");
+      _head = _tail = entry;
+    } else {
+      assert ((_tail->_next == NULL && _tail->_buffer != NULL), "Buffer queue is polluted");
+      _tail->_next = entry;
+      _tail = entry;
+    }
+    _length++;
+    assert(_tail->_next == NULL, "Bufer queue is polluted");
+  }
+
+  ParWriterBufferQueueElem* dequeue() {
+    if (_head == NULL)  return NULL;
+    ParWriterBufferQueueElem* entry = _head;
+    assert (entry->_buffer != NULL, "polluted buffer in writer list");
+    _head = entry->_next;
+    if (_head == NULL) {
+      _tail = NULL;
+    }
+    entry->_next = NULL;
+    _length--;
+    return entry;
+  }
+
+  bool is_empty() {
+    return _length == 0;
+  }
+
+  uint length() { return _length; }
+};
+
+// Support parallel heap dump.
+class ParDumpWriter : public AbstractDumpWriter {
+ private:
+  // Lock used to guarantee the integrity of multiple buffers writing.
+  static Monitor* _lock;
+  // Pointer of backend from global DumpWriter.
+  CompressionBackend* _backend_ptr;
+  char const * _err;
+  ParWriterBufferQueue* _buffer_queue;
+  size_t _internal_buffer_used;
+  char* _buffer_base;
+  bool _splited_data;
+  static const uint BackendFlushThreshold = 2;
+ protected:
+  virtual void flush(bool force = false) {
+    assert(_pos != 0, "must not be zero");
+    if (_pos != 0) {
+      refresh_buffer();
+    }
+
+    if (_splited_data || _is_huge_sub_record) {
+      return;
+    }
+
+    if (should_flush_buf_list(force)) {
+      assert(!_in_dump_segment && !_splited_data && !_is_huge_sub_record, "incomplete data send to backend!\n");
+      flush_to_backend(force);
+    }
+  }
+
+ public:
+  ParDumpWriter(DumpWriter* dw) :
```

Comment:
The DumpWriter constructor has the following comment, which you should add here also:
`// Check for error after constructing the object and destroy it in case of an error.`